src/main/java/rx/Single.java

```diff
@@ -685,6 +679,9 @@ public void call(SingleSubscriber<? super T> te) {
      * @see <a href="http://reactivex.io/documentation/operators/merge.html">ReactiveX operators documentation: Merge</a>
      */
     public final static <T> Single<T> merge(final Single<? extends Single<? extends T>> source) {
+        if (source.getClass() == ScalarSynchronousSingle.class) {
```

Comment:
It appears `instanceof` can be optimized much better by JIT than class equivalence checks. Could you rewrite this and all the other places?